### PR TITLE
Implementing changes in NanoAOD v12 & switching to mvaHZZIso for HZG electron Run 3

### PR DESCRIPTION
**Pull request by rzouCERN**
Added all the changes in variable type in nanoAOD v12.
Switched to mvaHZZIso for electrons for Run 3 (using WP Summer18UL (https://twiki.cern.ch/twiki/bin/viewauth/CMS/MultivariateElectronIdentificationRun2#HZZ_MVA_training_details_and_wor))
Updated btagging SFs for run 2 & 3. Added PNet btagging for Run 3.

### CODE REVIEW

File: src/process_nano.cxx
Comment (jdgrassi) [inline src/process_nano.cxx:1]: Line 222: comment labels are backwards: The section labeled for PNet and DeepCSV gives DeepJet working points, while the comment that says deepJet gives the PNet and deepCSV WPs.

File: txt/variables/nano
Comment (jdgrassi) [inline txt/variables/nano:1]: What is the reason to label some variables as 12.0, but others as just 12?

**Review by jdgrassi — COMMENTED** [review]
I'm not too familiar with how the variable naming (i.e. <array> name 12.0) works to create the desired behavior, so I have not commented on that.

File: src/process_nano.cxx
Comment (rzouCERN) [reply src/process_nano.cxx:1]: thanks! fixed in commit addf219

File: txt/variables/nano
Comment (rzouCERN) [reply txt/variables/nano:1]: They are the same. I changed them all to 12 for consistency.

File: src/el_producer.cpp
Comment (andersbarzdukas) [inline src/el_producer.cpp:1]: Not exactly about this PR. We have WPL for the Run-2 eras and HZZ WP for the Run 3 eras. Why is the default WP90? Should we add an error message if we ever use WP90 to let us know that the WP for electron ID is different than what we expect?

File: src/el_producer.cpp [hunk 1]
```diff
@@ -65,6 +65,14 @@ bool ElectronProducer::IsSignal(nano_tree &nano, int nano_idx, bool isZgamma) {
...
       case 2022:
       case 2023:
+	// got these values from : https://twiki.cern.ch/twiki/bin/viewauth/CMS/MultivariateElectronIdentificationRun2#HZZ_MVA_training_details_and_wor
+	//early Run 3 HZZ SF is approved applying the 2018 WP
+	return ((pt < 10. && fabs(eta) < 0.800 && nano.Electron_mvaHZZIso()[nano_idx] > 1.49603193295) ||\
+		(pt < 10. && fabs(eta) >= 0.800 && fabs(eta) < 1.479 && nano.Electron_mvaHZZIso()[nano_idx] > 1.52414154008) ||\
+		(pt < 10. && fabs(eta) >= 1.479 && nano.Electron_mvaHZZIso()[nano_idx] > 1.77694249574)||\
+		(pt >= 10. && fabs(eta) < 0.800 && nano.Electron_mvaHZZIso()[nano_idx] > 0.199463934736) ||\
+		(pt >= 10. && fabs(eta) >= 0.800 && fabs(eta) < 1.479 && nano.Electron_mvaHZZIso()[nano_idx] > 0.076063564084) ||\
+		(pt >= 10. && fabs(eta) >= 1.479 && nano.Electron_mvaHZZIso()[nano_idx] > -0.572118857519));
```
[the remark this answers is not in the document]
Comment (mcoshiro) [reply src/el_producer.cpp:75]: I did a small amount of CMSSW sleuthing, and based on some naming in [1] and [2], I think that the cuts provided may be for the raw value, which is related to the value stored in Nano by the equation

(nano)value = 2/(1+exp(-2*raw_value))-1

as seen in line 152 of [2]. However, I'm still not 100% certain.

[1] https://github.com/cms-sw/cmssw/blob/master/RecoEgamma/ElectronIdentification/python/Identification/mvaElectronID_Summer18UL_ID_ISO_cff.py
[2] https://github.com/cms-sw/cmssw/blob/774f85421329c42cd1d30ec43d470f527141fb92/RecoEgamma/EgammaTools/interface/MVAValueMapProducer.h

[the remark this answers is not in the document]
Comment (rzouCERN) [reply src/el_producer.cpp]: Thanks, Michael! I did some validation on this conversion on one nano sample (/net/cms11/cms11r0/pico/NanoAODv9/nano/2018/mc/ZGToLLG_01J_5f_TuneCP5_13TeV-amcatnloFXFX-pythia8__RunIISummer20UL18NanoAODv9__106X_upgrade2018_realistic_v16_L1v1-v1__130000__06D1490F-DB91-264B-A09E-8C66F6CA364E.root).

See example of the comparison below:
root [26] Events->Draw("Electron_mvaFall17V2Iso[0]", "Electron_mvaFall17V2Iso[0] < 0.95 && abs(Electron_eta[0]+Electron_deltaEtaSC[0]) < 0.8 && Electron_pt[0] < 10 &&  Electron_pt[0] > 5 && Electron_mvaFall17V2Iso[0] > 0.60477549249")
(long long) 327
root [27] Events->Draw("Electron_mvaFall17V2Iso[0]", "Electron_mvaFall17V2Iso[0] < 0.95 && abs(Electron_eta[0]+Electron_deltaEtaSC[0]) < 0.8 && Electron_pt[0] < 10 &&  Electron_pt[0] > 5 && Electron_mvaFall17V2Iso_WPL[0]==1")
(long long) 328
Taking the WPL from [https://github.com/cms-sw/cmssw/blob/master/RecoEgamma/ElectronIdentification/python/Identification/mvaElectronID_Fall17_iso_V2_cff.py](https://github.com/cms-sw/cmssw/blob/master/RecoEgamma/ElectronIdentification/python/Identification/mvaElectronID_Fall17_iso_V2_cff.py)

It looks like the conversion makes sense based on the WPL cuts in nano.

File: src/el_producer.cpp
Comment (rzouCERN) [reply src/el_producer.cpp:1]: I think it may just be that WP90 is a WP that exists in all eras? I added a warning message.

**Pull request merged**